sdc11073/mdib/transactions.py

```diff
@@ -0,0 +1,556 @@
+
+    def get_state_transaction_item(self, descriptor_handle):
+        """ if transaction has a state with given handle, return the transaction-item, otherwise None.  """
+        for lookup in (self.metric_state_updates,
+                       self.alert_state_updates,
+                       self.component_state_updates,
+                       self.context_state_updates,
+                       self.operational_state_updates,
+                       self.rt_sample_state_updates):
+            if descriptor_handle in lookup:
+                return lookup[descriptor_handle]
+        return None
+
+    @tr_method_wrapper
+    def add_state(self, state_container, adjust_state_version=True):
+        my_multi_key = self._device_mdib_container.states
+        my_updates = []
+        if state_container.isRealtimeSampleArrayMetricState:
+            my_updates = self.rt_sample_state_updates
+        elif state_container.isMetricState:
+            my_updates = self.metric_state_updates
+        elif state_container.isAlertState:
+            my_updates = self.alert_state_updates
+        elif state_container.isComponentState:
+            my_updates = self.component_state_updates
+        elif state_container.isSystemContextState:
+            my_updates = self.component_state_updates
+        elif state_container.isOperationalState:
+            my_updates = self.operational_state_updates
+        elif state_container.isContextState:
+            my_updates = self.context_state_updates
+            my_multi_key = self._device_mdib_container.context_states
+        elif state_container.NODETYPE == domTag('ScoState'):
+            # special case ScoState Draft6: cannot notify updates, it is a category of its own that does not fit anywhere
+            # This is a bug in the spec, not in this implementation!
+            return
+
+        descriptor_handle = state_container.descriptorHandle
+        my_handle = state_container.Handle if state_container.isContextState else state_container.descriptorHandle
+        if my_handle in my_updates:
+            raise ValueError(f'descriptorHandle {descriptor_handle} already in updated set!')
+
+        if adjust_state_version:
+            my_multi_key.set_version(state_container)
+        my_updates[descriptor_handle] = TrItem(None, state_container)  # old, new
+
+    def unget_state(self, state_container):
+        """ forget a state that was provided before by a getXXXState call"""
+        for lookup in (self.alert_state_updates, self.component_state_updates, self.context_state_updates,
+                       self.metric_state_updates, self.operational_state_updates, self.rt_sample_state_updates):
+            if state_container.descriptorHandle in lookup:
+                del lookup[state_container.descriptorHandle]
+
+    @tr_method_wrapper
+    def get_state(self, descriptor_handle, handle=None, adjust_state_version=True):

```

Comment:
reworked method so that is has no side effects. It only reads existing single states (not multi states), no implicit creation.
New methods get_context_state and mk_context_state handle cover the handling of multi states.